napari/layers/tracks/tracks.py

```diff
@@ -0,0 +1,435 @@
+# from napari.utils.events import Event
+# from napari.utils.colormaps import AVAILABLE_COLORMAPS
+
+from typing import Dict, List, Union
+
+import numpy as np
+
+from ...utils.colormaps import AVAILABLE_COLORMAPS
+from ...utils.events import Event
+from ..base import Layer
+from ._track_utils import TrackManager
+
+
+class Tracks(Layer):
+    """ Tracks
+
+    A Tracks layer for overlaying trajectories on image data.
+
+    Parameters
+    ----------
+    data : array (N, D)
+        Coordinates for N points in D dimensions. T(Z)YX
+    properties : dict {str: array (N,)}, DataFrame
+        Properties for each point. Each property should be an array of length N,
+        where N is the number of points. Must contain 'track_id'
+    graph : dict {int: list}
+        Graph representing track edges. Dictionary defines the mapping between
+        a track ID and the parents of the track. This can be one (the track
+        has one parent, and the parent has >=1 child) in the case of track
+        splitting, or more than one (the track has multiple parents, but
+        only one child) in the case of track merging.
+    color_by: str
+        track property (from property keys) to color vertices by
+    edge_width : float
+        Width for all vectors in pixels.
+    tail_length : float
+        Length of the projection of time as a tail, in units of time.
+    colormap : str
+        Default colormap to use to set vertex colors. Specialized colormaps,
+        relating to specified properties can be passed to the layer via
+        colormaps_dict.
+    colomaps_dict : dict {str: Colormap}
+        dictionary list of colormap objects to use for coloring by track
+        properties.
+    name : str
+        Name of the layer.
+    metadata : dict
+        Layer metadata.
+    scale : tuple of float
+        Scale factors for the layer.
+    translate : tuple of float
+        Translation values for the layer.
+    opacity : float
+        Opacity of the layer visual, between 0.0 and 1.0.
+    blending : str
+        One of a list of preset blending modes that determines how RGB and
+        alpha values of the layer visual get mixed. Allowed values are
+        {'opaque', 'translucent', and 'additive'}.
+    visible : bool
+        Whether the layer visual is currently being displayed.
+
+    """
+
+    # The max number of points that will ever be used to render the thumbnail
+    # If more points are present then they are randomly subsampled
+    _max_tracks_thumbnail = 1024
+
+    def __init__(
+        self,
+        data,
+        *,
+        properties=None,
+        graph=None,
+        edge_width=2,
+        tail_length=30,
+        n_dimensional=True,
+        name=None,
+        metadata=None,
+        scale=None,
+        translate=None,
+        opacity=1,
+        blending='additive',
+        visible=True,
+        colormap='viridis',
+        color_by='track_id',
+        colormaps_dict=None,
```

Comment:
I am not sure I understand the purpose of the `colormaps_dict` argument. Is it to provide custom colormaps? Or is it to support other ways to map color (e.g., color cycles or mappings of categorical property values to colors)?

Reply:
We've used this a lot to apply custom colormaps to properties, this includes color cycles and categorical properties. For example, cells may have a discrete state - it's useful to map the classification label to the track.

Reply:
Got it. Thanks for the explanation! I see how this functionality is really useful.

In other layers (e.g., points, shapes), we have the ability to color the points/shapes by colormaps and color cycles against the property values (see [here](https://github.com/napari/napari/blob/79500fe4ae0a8c3d0838d7330f4ac9a706b1bb9e/examples/add_points_with_properties.py#L29-L38)). However, explicit mappings between categorical values and colors isn't really supported in the public API. Although, it has been a requested feature (I'll try to dig up the issue later). I am mentioning this because I think whatever we decide here, we will want to (eventually) unify across the layers.

For the tracks layer, it sounds like at the minimum, we would want to support the following modes of coloring the tracks:
- `colormap` against a property with continuous values
- `color_cycle` against a property with categorical values. Here, I am defining a color cycle as a set of colors that gets assigned to the unique property values in a cyclic manner (i.e., if there are more unique values than colors, the colors get reused)
- `direct`: an explicit mapping of categorical property values to colors (e.g., a dictionary where the key is the property value and value is the color, see [the labels layer](https://github.com/napari/napari/blob/79500fe4ae0a8c3d0838d7330f4ac9a706b1bb9e/examples/add_labels_with_properties.py#L39-L54)). I think one of the neat things about the `colormaps_dict` is that you can pre-program in different modes for different properties!

Does that seem right, @quantumjot?

Eventually, I would like it if the tracks layer model handled all of these color modes internally (as opposed to accepting a custom colormap function). My rationale is that: (1) I think it opens the functionality up to more users and (2) it is easier to record the state of the viewer. I think there are a few paths forward:

1. we leave it as is
2. we make the API such that it can be used as is (i.e., users passes in a custom colormap function/dict) for now and it would not need to be changed when we add more of the color handling into the model in the future.
3. we add the color modes into the model now

I am leaning towards option (2) as it seems like the right middle ground. Although, I'm not exactly sure how the API would look yet. While option (3) is the most work, we do have code for `colormap` and `color_cycle` modes already in the points and shapes layers and `direct` mode in the labels layer.

Sorry to harp on this point so much. It's just that if we include this as a kwarg in the public API, I think it could be hard to change in the future because people will be depending on it. In any case, I am curious about what everyone thinks!

Reply:
Yeah - I think those three modes cover all the use cases I can think of!

BTW, I implemented a `color_cycle` using a subclass of the new `Colormap` object by adding `np.mod(values, len(self.controls))` to the `map` function - seems to work quite nicely.

Reply:
> I think one of the neat things about the colormaps_dict is that you can pre-program in different modes for different properties!

Exactly! That is super useful!!